File: cpp_solutions/2406. Divide Intervals Into Minimum Number of Groups.cpp
class Solution {
public:
    // Simple Line Sweep Algorithm
    void PrintMap(map<float, int> arr) {
        
        for (auto it : arr){
            cout << it.first << "--" << it.second << "\n";
        }
    }
    
    void PrintVec(vector<vector<float>> arr) {
        
        for (auto it : arr){
            for(auto its : it) {
                cout << its << "--";
            }
            cout << "\n";
        }
    }
    
    int minGroups(vector<vector<int>>& intervals) {
        
        map<float, int> dicts;
        
        for(auto it : intervals) {
            float a = it[0] - 0.01;
            if (dicts.find(a)!=dicts.end()) {
                dicts[a]+=1;
            }  else {
                dicts[a] = 1;
            }
            
            
            float b = float(it[1]);
            if (dicts.find(b)!=dicts.end()) {
                dicts[b]-=1;
            }  else {
                dicts[b] = -1;
            }
        }
        
        
        vector<vector<float>> arr;
        
        for (auto it : dicts) {
            vector<float> temp;
            temp.emplace_back(it.first);
            temp.emplace_back(it.second);
            arr.emplace_back(temp);
        }
        
        // PrintVec(arr);
        sort(arr.begin(), arr.end());
        // PrintVec(arr);
        int curr = 0;
        int ans = 0;
        
        for(auto it : arr) {
            curr+= int(it[1]);
            ans = max(ans, curr);
        }
        return ans;
    }
};

File: python3_solution_set2/2406. Divide Intervals Into Minimum Number of Groups.py
class Solution:
    def minGroups(self, intervals: List[List[int]]) -> int:
        
        ## Simple line sweep algorithm
        ## Time Complexity : O(nlogn)
        ## Space Complexity : O(n)
        
        intervals = sorted(intervals, key = lambda x:x[0])
        dicts = {}
        
        ## where an interval starts overlap+=1
        ## where an interval ends overlap-=1
        ## since [1,5] and [5,8] is an overlap
        ## we add [1,4.99] and [5,7.99] to avoid this miscalculation
        
        ## max overlap at any given point will be the number of groups required
        for x,y in intervals:
            a = x - 0.01
            if a in dicts:
                dicts[a]+=1
                
            else:
                dicts[a] = 1
                
            if y in dicts:
                dicts[y]-=1
            else:
                dicts[y] = -1
                
            
        t = []
        
        for x in dicts:
            t.append([x,dicts[x]])
            
        t = sorted(t, key = lambda x:x[0])
        
        ans = 0
        curr = 0
        for x,y in t:
            curr+=y
            ans = max(ans, curr)
            
        return ans
                
            
            
        
